File: app/src/main/java/com/example/agrinova/DiseaseDetectionActivity.java
package com.example.agrinova;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;

import com.example.agrinova.databinding.ActivityDiseaseDetectionBinding;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class DiseaseDetectionActivity extends AppCompatActivity {

    private static final String TAG = "DiseaseDetection";
    private static final int PERMISSION_REQUEST_CODE = 1001;
    private static final String DATABASE_URL = "https://agrinova-9236c-default-rtdb.asia-southeast1.firebasedatabase.app/";

    private ActivityDiseaseDetectionBinding binding;
    private TFLiteHelper tfliteHelper;
    private Uri currentPhotoUri;
    private String currentPhotoPath;
    private String lastDetectedDisease = "None";
    private String lastDetectedCrop = "Unknown";

    // Modern Activity Result Launchers
    private final ActivityResultLauncher<String> galleryLauncher = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    processImage(uri);
                }
            });

    private final ActivityResultLauncher<Uri> cameraLauncher = registerForActivityResult(
            new ActivityResultContracts.TakePicture(),
            isSuccess -> {
                if (isSuccess && currentPhotoUri != null) {
                    processImage(currentPhotoUri);
                } else {
                    Log.d(TAG, "Camera capture cancelled or failed");
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            binding = ActivityDiseaseDetectionBinding.inflate(getLayoutInflater());
            setContentView(binding.getRoot());

            setSupportActionBar(binding.toolbar);
            if (getSupportActionBar() != null) {
                getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            }
            binding.toolbar.setNavigationOnClickListener(v -> finish());

            initModel();
            setupClickListeners();

            if (savedInstanceState != null) {
                currentPhotoPath = savedInstanceState.getString("photo_path");
                if (currentPhotoPath != null) {
                    currentPhotoUri = Uri.fromFile(new File(currentPhotoPath));
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Initialization failed", e);
            finish();
        }
    }

    private void initModel() {
        try {
            tfliteHelper = new TFLiteHelper(this);
            boolean isInitialized = tfliteHelper.init("crop_disease_model.tflite", "labels.txt");
            if (!isInitialized) {
                Toast.makeText(this, "Failed to load detection model", Toast.LENGTH_LONG).show();
            }
        } catch (Exception e) {
            Log.e(TAG, "Model initialization failed", e);
        }
    }

    private void setupClickListeners() {
        binding.btnCamera.setOnClickListener(v -> checkPermissionsAndLaunch(true));
        binding.btnGallery.setOnClickListener(v -> checkPermissionsAndLaunch(false));
    }

    private void checkPermissionsAndLaunch(boolean isCamera) {
        String[] permissions;
        if (isCamera) {
            permissions = new String[]{Manifest.permission.CAMERA};
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                permissions = new String[]{Manifest.permission.READ_MEDIA_IMAGES};
            } else {
                permissions = new String[]{Manifest.permission.READ_EXTERNAL_STORAGE};
            }
        }

        boolean allGranted = true;
        for (String p : permissions) {
            if (ContextCompat.checkSelfPermission(this, p) != PackageManager.PERMISSION_GRANTED) {
                allGranted = false;
                break;
            }
        }

        if (allGranted) {
            if (isCamera) launchCamera(); else galleryLauncher.launch("image/*");
        } else {
            ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if (permissions[0].equals(Manifest.permission.CAMERA)) {
                    launchCamera();
                } else {
                    galleryLauncher.launch("image/*");
                }
            } else {
                Toast.makeText(this, "Permission required to use this feature", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void launchCamera() {
        try {
            File photoFile = createImageFile();
            if (photoFile != null) {
                currentPhotoUri = FileProvider.getUriForFile(this, getPackageName() + ".fileprovider", photoFile);
                cameraLauncher.launch(currentPhotoUri);
            }
        } catch (IOException ex) {
            Log.e(TAG, "Camera launch failed", ex);
            Toast.makeText(this, "Could not create image file", Toast.LENGTH_SHORT).show();
        }
    }

    private File createImageFile() throws IOException {
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(new Date());
        File storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        File image = File.createTempFile("DET_" + timeStamp, ".jpg", storageDir);
        currentPhotoPath = image.getAbsolutePath();
        return image;
    }

    private void processImage(Uri imageUri) {
        if (imageUri == null) return;
        
        binding.loadingProgress.setVisibility(View.VISIBLE);
        binding.tvPlaceholder.setVisibility(View.GONE);
        binding.layoutResults.setVisibility(View.GONE);

        new Thread(() -> {
            try {
                Bitmap bitmap = loadResizedBitmap(imageUri);
                if (bitmap == null) {
                    runOnUiThread(() -> {
                        binding.loadingProgress.setVisibility(View.GONE);
                        Toast.makeText(this, "Failed to load image", Toast.LENGTH_SHORT).show();
                    });
                    return;
                }

                List<TFLiteHelper.Recognition> results = tfliteHelper.runInference(bitmap);

                runOnUiThread(() -> {
                    binding.loadingProgress.setVisibility(View.GONE);
                    binding.ivPreview.setAlpha(1.0f);
                    binding.ivPreview.setImageBitmap(bitmap);
                    
                    if (results != null && !results.isEmpty()) {
                        displayResults(results);
                    } else {
                        Toast.makeText(this, "No disease detected. Try a clearer photo.", Toast.LENGTH_LONG).show();
                    }
                });
            } catch (Exception e) {
                Log.e(TAG, "Processing error", e);
                runOnUiThread(() -> {
                    binding.loadingProgress.setVisibility(View.GONE);
                    Toast.makeText(this, "Error processing image", Toast.LENGTH_SHORT).show();
                });
            }
        }).start();
    }

    private Bitmap loadResizedBitmap(Uri uri) throws IOException {
        InputStream inputStream = getContentResolver().openInputStream(uri);
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeStream(inputStream, null, options);
        if (inputStream != null) inputStream.close();

        int targetSize = 1024;
        int scaleFactor = Math.min(options.outWidth / targetSize, options.outHeight / targetSize);

        options.inJustDecodeBounds = false;
        options.inSampleSize = Math.max(1, scaleFactor);
        
        inputStream = getContentResolver().openInputStream(uri);
        Bitmap bitmap = BitmapFactory.decodeStream(inputStream, null, options);
        if (inputStream != null) inputStream.close();
        return bitmap;
    }

    private void displayResults(List<TFLiteHelper.Recognition> results) {
        TFLiteHelper.Recognition top = results.get(0);
        lastDetectedCrop = top.getCropName();
        lastDetectedDisease = top.getDiseaseName();
        
        binding.layoutResults.setVisibility(View.VISIBLE);
        binding.tvCropName.setText("Crop: " + lastDetectedCrop);
        binding.tvDiseaseName.setText("Disease: " + lastDetectedDisease);
        float confidenceValue = top.getConfidence() * 100;
        binding.tvConfidence.setText(String.format(Locale.getDefault(), "Confidence: %.1f%%", confidenceValue));

        if (results.size() > 1) {
            StringBuilder others = new StringBuilder("Other possibilities: ");
            for (int i = 1; i < results.size(); i++) {
                others.append(results.get(i).getDiseaseName()).append(" (")
                      .append(String.format(Locale.getDefault(), "%.1f%%", results.get(i).getConfidence() * 100))
                      .append(")");
                if (i < results.size() - 1) others.append(", ");
            }
            binding.tvTopPredictions.setText(others.toString());
        } else {
            binding.tvTopPredictions.setText("");
        }

        updateRecommendations(top.getTitle());
        saveToFirebase(lastDetectedCrop, lastDetectedDisease, String.format(Locale.getDefault(), "%.0f%%", confidenceValue));
    }

    private void updateRecommendations(String label) {
        Map<String, String[]> recommendations = new HashMap<>();
        recommendations.put("Apple___Apple_scab", new String[]{
            "Apply fungicides like Captan or Mancozeb. Rake and destroy fallen leaves.",
            "Neem oil spray can be effective in early stages.",
            "Water at the base of the tree to keep leaves dry. Plant resistant varieties."
        });
        // ... (rest of recommendations)
        
        String[] advice = recommendations.get(label);
        String treatment, organic, prevention;
        if (advice == null) {
            if (label.toLowerCase().contains("healthy")) {
                treatment = "Your crop looks healthy! Continue regular maintenance.";
                organic = "Use organic compost for better yield.";
                prevention = "Keep monitoring for any signs of pests or changes.";
            } else {
                treatment = "Consult a local agricultural expert. Generic antifungal spray might help.";
                organic = "Neem oil is a broad-spectrum organic solution.";
                prevention = "Ensure proper drainage and avoid overcrowding plants.";
            }
        } else {
            treatment = advice[0];
            organic = advice[1];
            prevention = advice[2];
        }
        
        binding.tvTreatment.setText(treatment);
        binding.tvOrganic.setText(organic);
        binding.tvPrevention.setText(prevention);
    }

    private void saveToFirebase(String crop, String disease, String percentage) {
        try {
            FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
            if (user == null) return;

            DatabaseReference ref = FirebaseDatabase.getInstance(DATABASE_URL).getReference("CropQuality");
            String id = ref.push().getKey();
            String timestamp = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault()).format(new Date());

            CropQuality qualityData = new CropQuality(
                id,
                crop,
                disease.contains("Healthy") ? "A Grade" : "B Grade", // Dynamic grade based on health
                "N/A", // Price not calculated here
                percentage,
                disease,
                binding.tvTreatment.getText().toString(),
                user.getEmail(),
                user.getUid(),
                timestamp
            );

            if (id != null) {
                ref.child(id).setValue(qualityData.toMap())
                    .addOnFailureListener(e -> Log.e(TAG, "Firebase save failed", e));
            }
        } catch (Exception e) {
            Log.e(TAG, "Firebase operation failed", e);
        }
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("photo_path", currentPhotoPath);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (tfliteHelper != null) {
            tfliteHelper.close();
        }
    }
}

File: app/src/main/java/com/example/agriguard/activities/DiseaseDetectionActivity.kt
package com.example.agriguard.activities

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.lifecycle.ViewModelProvider
import com.example.agriguard.R
import com.example.agriguard.databinding.ActivityDiseaseDetectionBinding
import com.example.agriguard.models.PredictionResult
import com.example.agriguard.utils.BitmapUtils
import com.example.agriguard.viewmodels.DetectionViewModel
import android.util.Log
import com.example.agriguard.BuildConfig
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

class DiseaseDetectionActivity : AppCompatActivity() {

    private lateinit var binding: ActivityDiseaseDetectionBinding
    private lateinit var viewModel: DetectionViewModel
    private var currentImageUri: Uri? = null

    private val takePictureLauncher = registerForActivityResult(
        ActivityResultContracts.TakePicture()
    ) { success: Boolean ->
        if (success) {
            currentImageUri?.let { showPreview(it) }
        }
    }

    private val galleryLauncher = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { uri: Uri? ->
        uri?.let { showPreview(it) }
    }

    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted: Boolean ->
        if (isGranted) {
            launchCamera()
        } else {
            Toast.makeText(this, "Camera permission denied", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDiseaseDetectionBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(this).get(DetectionViewModel::class.java)

        if (BuildConfig.PLANT_ID_API_KEY.isEmpty()) {
            Toast.makeText(this, "Plant.id API Key is missing!", Toast.LENGTH_LONG).show()
        }

        setupListeners()
        observeViewModel()
    }

    private fun setupListeners() {
        binding.toolbar.setNavigationOnClickListener { finish() }
        binding.btnCapture.setOnClickListener { checkCameraPermission() }
        binding.btnGallery.setOnClickListener { galleryLauncher.launch("image/*") }
        binding.btnDetect.setOnClickListener { startDetection() }
        binding.btnRemoveImage.setOnClickListener { resetUI() }
        binding.btnTryAgain.setOnClickListener { resetUI() }
    }

    private fun observeViewModel() {
        viewModel.isLoading.observe(this) { loading: Boolean ->
            binding.progressBar.visibility = if (loading) View.VISIBLE else View.GONE
            binding.btnDetect.isEnabled = !loading
            if (loading) {
                binding.resultLayout.visibility = View.GONE
                binding.errorCard.visibility = View.GONE
            }
        }

        viewModel.error.observe(this) { error: String? ->
            if (error != null) {
                showError("❌ Analysis Error", error)
            }
        }

        viewModel.result.observe(this) { result: PredictionResult? ->
            if (result != null) {
                if (result.status == "Invalid") {
                    showError("❌ Invalid Crop Image", result.description)
                } else {
                    displayResult(result)
                }
            }
        }
    }

    private fun showError(title: String, message: String) {
        binding.errorCard.visibility = View.VISIBLE
        binding.resultLayout.visibility = View.GONE
        binding.tvErrorTitle.text = title
        binding.tvErrorMessage.text = message
    }

    private fun displayResult(result: PredictionResult) {
        // Requirement 3 & 5: Clear previous state and show layout
        binding.resultLayout.visibility = View.VISIBLE
        binding.errorCard.visibility = View.GONE

        val isHealthy = result.status.equals("Healthy", ignoreCase = true)

        binding.tvResultStatus.text = result.status
        binding.tvConfidence.text = result.getConfidenceString()
        
        // Requirement 2 & 3: Handle Unknown Crop and Warning Badge
        if (!result.isCropIdentified) {
            binding.tvResultCrop.text = "Crop: Unknown Crop"
            binding.tvCropWarning.visibility = View.VISIBLE
        } else {
            binding.tvResultCrop.text = "Crop: ${result.cropName}"
            binding.tvCropWarning.visibility = View.GONE
        }

        binding.tvResultDisease.text = if (isHealthy) "None" else result.diseaseName
        
        // UI Styling based on status
        if (isHealthy) {
            binding.tvResultStatus.setBackgroundResource(R.drawable.bg_status_healthy)
            binding.tvResultDisease.setTextColor(ContextCompat.getColor(this, R.color.primary_green))
        } else {
            binding.tvResultStatus.setBackgroundResource(R.drawable.bg_status_diseased)
            binding.tvResultDisease.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_dark))
        }

        // Requirement 1: ALWAYS show details, even if healthy
        binding.cardDetails.visibility = View.VISIBLE
        binding.cardRecommendations.visibility = View.VISIBLE
        binding.cardPrevention.visibility = View.VISIBLE
        
        // Populating text fields
        binding.tvSymptoms.text = result.symptoms
        binding.tvCauses.text = result.causes
        binding.tvFertilizers.text = result.fertilizer
        binding.tvPesticides.text = result.pesticides
        binding.tvPrevention.text = result.preventionTips
        binding.tvIrrigation.text = result.irrigationTips
    }

    private fun startDetection() {
        currentImageUri?.let { uri ->
            try {
                val bitmap = BitmapUtils.getCorrectlyOrientedBitmap(this, uri)
                viewModel.detect(bitmap)
            } catch (e: IOException) {
                Toast.makeText(this, "Failed to process image", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun showPreview(uri: Uri) {
        currentImageUri = uri
        try {
            val bitmap = BitmapUtils.getCorrectlyOrientedBitmap(this, uri)
            binding.ivPreview.setImageBitmap(bitmap)
        } catch (e: IOException) {
            binding.ivPreview.setImageURI(uri)
        }
        binding.ivPreview.visibility = View.VISIBLE
        binding.btnRemoveImage.visibility = View.VISIBLE
        binding.uploadPlaceholder.visibility = View.GONE
        binding.btnDetect.visibility = View.VISIBLE
        binding.resultLayout.visibility = View.GONE
        binding.errorCard.visibility = View.GONE
    }

    private fun resetUI() {
        currentImageUri = null
        binding.ivPreview.setImageURI(null)
        binding.ivPreview.visibility = View.GONE
        binding.btnRemoveImage.visibility = View.GONE
        binding.uploadPlaceholder.visibility = View.VISIBLE
        binding.btnDetect.visibility = View.GONE
        binding.resultLayout.visibility = View.GONE
        binding.errorCard.visibility = View.GONE
    }

    private fun checkCameraPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
            launchCamera()
        } else {
            requestPermissionLauncher.launch(Manifest.permission.CAMERA)
        }
    }

    private fun launchCamera() {
        try {
            val photoFile = createImageFile()
            val uri = FileProvider.getUriForFile(
                this,
                "com.example.agriguard.fileprovider",
                photoFile
            )
            currentImageUri = uri
            takePictureLauncher.launch(uri)
        } catch (ex: IOException) {
            Toast.makeText(this, "Error creating file", Toast.LENGTH_SHORT).show()
        }
    }

    @Throws(IOException::class)
    private fun createImageFile(): File {
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        val storageDir = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        return File.createTempFile("JPEG_${timeStamp}_", ".jpg", storageDir)
    }
}
